_posts/2017-03-28-bettracker-typescript.md

```diff
@@ -0,0 +1,116 @@
+---
+layout:     post
+title:      Modernizing our Application with TypeScript
+date:       2017-03-28 09:00
+summary:    How we used TypeScript to successfully refactor a large business critical codebase.
+category:   Software Engineering
+tags:       TypeScript, react, re-factoring
+author:     james_moorhouse
+---
+
+## Introduction
+
+Back in October 2016 my team were looking for ways we could improve our My Bets product which was written using ReactJS. We'd first made the product about 2 and a half years ago, much before my time, when ReactJS was still fairly new on the scene. As such the ReactJS itself has been through quite a few changes since then and a few of our design patterns were in need of a spring clean.
+
+As a product, My Bets allows our customers to view, track and cash out any of their bets. Bets these days are complex: they can consist of many selections with multiple outcomes across multiple events; then we need to track the history of a bet - i.e. all the data at the time it was placed compared to the equivalent, current data.
+
+![My Bets Expanded](/images/mybets-typescript/mybets-expanded.png)
+
+Confused? don't worry, you're not alone. The important thing to note is that we needed to be able to model complex data structures and their relationships to each other. TypeScript seemed perfect for this. Like using babel, it promised to give us ES6 features ahead of time but also allowed us to add strict types to give us more control over our data structure.
+
+## What is TypeScript
+If you're not familiar with [TypeScript](https://github.com/Microsoft/TypeScript/), it's an open source language created by Microsoft  which compiles down into vanilla JavaScript. JavaScript is itself TypeScript, TypeScript is just a superset of javascript which includes typings and a few other bits of syntactic sugar such as Enums and Interfaces - things that you'd be used to seeing in "grown up" OOP languages.
```

Comment:
Inconsistent casing on "JavaScript" (there's one instance where it's all lowercase).